### DISCUSSION TITLE
MASM0010: Unable to unmarshall metro config file from location

### DISCUSSION THREAD

**Discussion by rjkroll** [nextgenhealthcare/connect#5747] [Q&A] [answered]
Hi,

We recently upgraded from Mirth 3.x to Mirth 4.2 (using OpenJDK 19.0.2).   We have a channel is has a Webservice Sender as the Destination, which was in existence with the Mirth 3.x usage.   After the upgrade, it seems to work normally for awhile, but every month or so, all the messages start queuing up and we receive the same error shown below over and over.   Restarting the Mirth service solves the problem and the queued messages are sent and everything is back to normal for another few weeks.

I have never seen this type of error before?  Can anyone shed some light as to what it means and how it might be resolved?

`[2023-04-05 11:09:10.848]  ERROR  (com.sun.metro.assembler:27): MASM0010: Unable to unmarshall metro config file from location [ jar:file:/C:/Program%20Files/Mirth%20Connect/server-lib/javax/jaxws/jaxws-rt-2.3.0.2.jar!/META-INF/jaxws-tubes-default.xml ]
javax.xml.bind.JAXBException: Implementation of JAXB-API has not been found on module path or classpath. - with linked exception:[java.lang.ClassNotFoundException: com.sun.xml.internal.bind.v2.ContextFactory]
	at javax.xml.bind.ContextFinder.newInstance(ContextFinder.java:177)
	at javax.xml.bind.ContextFinder.find(ContextFinder.java:341)
	at javax.xml.bind.JAXBContext.newInstance(JAXBContext.java:508)
	at javax.xml.bind.JAXBContext.newInstance(JAXBContext.java:465)
	at javax.xml.bind.JAXBContext.newInstance(JAXBContext.java:366)
	at com.sun.xml.ws.assembler.MetroConfigLoader.createJAXBContext(MetroConfigLoader.java:298)
	at com.sun.xml.ws.assembler.MetroConfigLoader.loadMetroConfig(MetroConfigLoader.java:256)
	at com.sun.xml.ws.assembler.MetroConfigLoader.init(MetroConfigLoader.java:147)
	at com.sun.xml.ws.assembler.MetroConfigLoader.<init>(MetroConfigLoader.java:121)
	at com.sun.xml.ws.assembler.TubelineAssemblyController.getTubeCreators(TubelineAssemblyController.java:93)
	at com.sun.xml.ws.assembler.MetroTubelineAssembler.createClient(MetroTubelineAssembler.java:118)
	... 29 more`

**Comment by jonbartels** [comment]
Interesting problem.

A workaround, though probably a lot of effort, would be to use an HTTP sender instead of the WS Sender. How complicated is the WSDL you're invoking?

This is unusual because it would have to be able to read the file on startup. Further that file is a file contained on the classpath inside a JAR. It's not like a file somewhere on the filesystem that could be moved or have permissions changed. It should just be there all the time.

Are there any errors PRIOR to the error you showed? 

I did some scrounging and found a few references:
* https://stackoverflow.com/questions/54201296/jax-ws-call-does-not-work-when-calling-java-from-javascript-using-jxbrowser - this is interesting because its Java called from JS. It suggests a certain config property, but the link to details is dead
* https://github.com/kivy/pyjnius/issues/582 - Is also interesting. It suggests that new thread cannot read the file when an old one could. For your case in Mirth perhaps Mirth is spawning a new thread. 
* https://www.ibm.com/docs/en/configurepricequote/10.0?topic=SS4QMC_10.0.0/troubleshooting/ts_DeployEARbuiltwihJAXWSwebservice.htm - This calls out a workaround that might actually work. Unpack the JAR (it's a ZIP file), copy the `jaxws-tubes-default.xml` file, save that file to a resource directory, add that resource directory to mirth, make your channel use that directory. Then it _might_ pick up the file from the directory which is now on the classpath

I'm making a lot of guesses here but this issue sounds challenging and I did not want you to go without a reply. The general strategy you want to apply here is to find how to solve this in pure Java, then translate that Java solution to one that works in Mirth and its classloading patterns.

**Reply by jonbartels** [reply]
Oh another idea - you upgraded from Mirth 3.x to Mirth 4.2, did any of the JAX-WS or SOAP or Metro related JARs change between those versions?

Seems like no - https://github.com/nextgenhealthcare/connect/tree/5a2e912a2a7f39ad01c45ec1801714d13117b11d/server/lib/javax/jaxws no updates in 5 years!

I wonder if the older libraries are compatible with Java 19

**Reply by pacmano1** [reply]
Starting from Java 9, JAXB was been removed from the core Java runtime and is now provided separately.

**Reply by jonbartels** [reply, marked as answer]
@rjkroll ignore me and read @pacmano1 answer. He's right.

You need to acquire the JAXB JARs for JAX-WS and put those into a resource directory then add that directory as a channel dependency.

**Reply by pacmano1** [reply]
I am confused with "it seems to work normally for awhile" though.

**Reply by jonbartels** [reply]
What Java version were you using with MC 3.x @rjkroll ?

**Reply by rjkroll** [reply]
HI,

Thank you for the detailed responses.  

@pacmano1 "It works normally for awhile" means that the interface will send the messages without any issues.  No alerts, no errors in the log, nothing.     We upgraded to Mirth 4.2 and OpenJDK (from Mirth 3.x and Java 8) in January is the channel worked fine until the end of Feb, where this exact error first was encountered.   The messages then just started queueing up until I restarted the Mirth server.    Then, messages were sent fine again until this week, where the same error happened and messages queued again until the service was restarted.

I will start looking into obtaining or adding the JAXB/JAX-WS rars and seeing how I can get them into Mirth.   Maybe this is happening because we did upgrades instead of fresh installs perhaps or its permissions based somehow?   Really at a loss, as this is the only time i've seen this in the last 10 years of using MC.

**Reply by pacmano1** [reply]
I just meant I would think all calls would fail if the library was missing.
